File: crates/kuben-store/src/repo/tokens.rs
//! API tokens (scenario 3). Only `sha256(secret)` is stored; the token id is
//! public and used for the lookup, the secret is compared in constant time by
//! the API layer.

use kuben_core::{
    ids::{OrgId, TokenId, UserId},
    model::{ApiToken, TokenScope},
    time::now_ms,
};

use crate::{
    Store, StoreError,
    db::{with_reader, with_writer},
};

/// Input for a new token.
#[derive(Debug)]
pub struct NewToken {
    pub id: TokenId,
    pub org_id: OrgId,
    pub owner: UserId,
    pub name: String,
    pub prefix: String,
    pub secret_hash: Vec<u8>,
    pub scope: TokenScope,
    pub expires_at: Option<i64>,
}

#[derive(Debug, sqlx::FromRow)]
struct TokenRow {
    id: String,
    org_id: String,
    owner_user_id: Option<String>,
    name: String,
    prefix: String,
    secret_hash: Vec<u8>,
    scopes: String,
    expires_at: Option<i64>,
    last_used_at: Option<i64>,
    revoked_at: Option<i64>,
    created_at: i64,
}

impl TryFrom<TokenRow> for ApiToken {
    type Error = StoreError;
    fn try_from(r: TokenRow) -> Result<Self, Self::Error> {
        let bad_id = |e: uuid::Error| sqlx::Error::Decode(e.into());
        Ok(Self {
            id: r.id.parse().map_err(bad_id)?,
            org_id: r.org_id.parse().map_err(bad_id)?,
            owner: r.owner_user_id.map(|o| o.parse().map_err(bad_id)).transpose()?,
            name: r.name,
            prefix: r.prefix,
            secret_hash: r.secret_hash,
            scope: serde_json::from_str(&r.scopes).map_err(|e| sqlx::Error::Decode(e.into()))?,
            expires_at: r.expires_at,
            last_used_at: r.last_used_at,
            revoked_at: r.revoked_at,
            created_at: r.created_at,
        })
    }
}

const INSERT_TOKEN: &str = "INSERT INTO api_tokens \
     (id, org_id, owner_user_id, name, prefix, secret_hash, scopes, expires_at, created_at) \
     VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)";
const SELECT_TOKEN: &str = "SELECT id, org_id, owner_user_id, name, prefix, secret_hash, scopes, expires_at, \
     last_used_at, revoked_at, created_at FROM api_tokens WHERE id = $1";
const SELECT_TOKENS_OF_OWNER: &str = "SELECT id, org_id, owner_user_id, name, prefix, secret_hash, scopes, \
     expires_at, last_used_at, revoked_at, created_at FROM api_tokens WHERE owner_user_id = $1 \
     ORDER BY created_at DESC";
const REVOKE_TOKEN: &str =
    "UPDATE api_tokens SET revoked_at = $3 WHERE id = $1 AND owner_user_id = $2 AND revoked_at IS NULL";
const REVOKE_USER_TOKENS: &str =
    "UPDATE api_tokens SET revoked_at = $3 WHERE org_id = $1 AND owner_user_id = $2 AND revoked_at IS NULL";
const TOUCH_TOKEN: &str = "UPDATE api_tokens SET last_used_at = $2 WHERE id = $1";

impl Store {
    pub async fn create_token(&self, t: NewToken) -> Result<ApiToken, StoreError> {
        let created_at = now_ms();
        let scopes = serde_json::to_string(&t.scope).map_err(|e| sqlx::Error::Encode(e.into()))?;
        with_writer!(self, |pool| {
            sqlx::query(INSERT_TOKEN)
                .bind(t.id.to_string())
                .bind(t.org_id.to_string())
                .bind(t.owner.to_string())
                .bind(&t.name)
                .bind(&t.prefix)
                .bind(&t.secret_hash)
                .bind(&scopes)
                .bind(t.expires_at)
                .bind(created_at)
                .execute(pool)
                .await?;
        });
        Ok(ApiToken {
            id: t.id,
            org_id: t.org_id,
            owner: Some(t.owner),
            name: t.name,
            prefix: t.prefix,
            secret_hash: t.secret_hash,
            scope: t.scope,
            expires_at: t.expires_at,
            last_used_at: None,
            revoked_at: None,
            created_at,
        })
    }

    pub async fn find_token(&self, id: TokenId) -> Result<Option<ApiToken>, StoreError> {
        let row: Option<TokenRow> = with_reader!(self, |pool| sqlx::query_as(SELECT_TOKEN)
            .bind(id.to_string())
            .fetch_optional(pool)
            .await?);
        row.map(ApiToken::try_from).transpose()
    }

    pub async fn list_tokens(&self, owner: UserId) -> Result<Vec<ApiToken>, StoreError> {
        let rows: Vec<TokenRow> = with_reader!(self, |pool| sqlx::query_as(SELECT_TOKENS_OF_OWNER)
            .bind(owner.to_string())
            .fetch_all(pool)
            .await?);
        rows.into_iter().map(ApiToken::try_from).collect()
    }

    /// Revoke one of `owner`'s tokens. `false` when there was nothing to revoke.
    pub async fn revoke_token(&self, id: TokenId, owner: UserId) -> Result<bool, StoreError> {
        let n = with_writer!(self, |pool| sqlx::query(REVOKE_TOKEN)
            .bind(id.to_string())
            .bind(owner.to_string())
            .bind(now_ms())
            .execute(pool)
            .await?
            .rows_affected());
        Ok(n > 0)
    }

    /// Revoke every token a user owns in an org (used when a member is removed).
    pub async fn revoke_user_tokens(&self, org: OrgId, user: UserId) -> Result<u64, StoreError> {
        let n = with_writer!(self, |pool| sqlx::query(REVOKE_USER_TOKENS)
            .bind(org.to_string())
            .bind(user.to_string())
            .bind(now_ms())
            .execute(pool)
            .await?
            .rows_affected());
        Ok(n)
    }

    pub async fn touch_token(&self, id: TokenId) -> Result<(), StoreError> {
        with_writer!(self, |pool| {
            sqlx::query(TOUCH_TOKEN)
                .bind(id.to_string())
                .bind(now_ms())
                .execute(pool)
                .await?;
        });
        Ok(())
    }
}

File: internal/store/tokens.go
package store

// API tokens (scenario 3); the port of repo/tokens.rs. Only sha256(secret)
// is stored; the token id is public and used for the lookup, the secret is
// compared in constant time by the API layer.

import (
	"context"
	"encoding/json"

	"github.com/jackc/pgx/v5"

	"github.com/Teamtem-dev/kuben/internal/core/ids"
	"github.com/Teamtem-dev/kuben/internal/core/model"
	"github.com/Teamtem-dev/kuben/internal/core/opt"
)

// NewToken is the input for a new token.
type NewToken struct {
	ID         ids.TokenID
	OrgID      ids.OrgID
	Owner      ids.UserID
	Name       string
	Prefix     string
	SecretHash []byte
	Scope      model.TokenScope
	ExpiresAt  opt.Val[int64]
}

const (
	insertToken = "INSERT INTO api_tokens " +
		"(id, org_id, owner_user_id, name, prefix, secret_hash, scopes, expires_at, created_at) " +
		"VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)"
	selectToken = "SELECT id, org_id, owner_user_id, name, prefix, secret_hash, scopes, expires_at, " +
		"last_used_at, revoked_at, created_at FROM api_tokens WHERE id = $1"
	selectTokensOfOwner = "SELECT id, org_id, owner_user_id, name, prefix, secret_hash, scopes, " +
		"expires_at, last_used_at, revoked_at, created_at FROM api_tokens " +
		"WHERE owner_user_id = $1 AND ci_policy_id IS NULL ORDER BY created_at DESC"
	revokeToken      = "UPDATE api_tokens SET revoked_at = $3 WHERE id = $1 AND owner_user_id = $2 AND revoked_at IS NULL"
	revokeUserTokens = "UPDATE api_tokens SET revoked_at = $3 WHERE org_id = $1 AND owner_user_id = $2 AND revoked_at IS NULL"
	touchToken       = "UPDATE api_tokens SET last_used_at = $2 WHERE id = $1"
)

func scanToken(row pgx.CollectableRow) (model.APIToken, error) {
	var t model.APIToken
	var owner *ids.UserID
	var scopes string
	var expires, lastUsed, revoked *int64
	err := row.Scan(&t.ID, &t.OrgID, &owner, &t.Name, &t.Prefix, &t.SecretHash, &scopes,
		&expires, &lastUsed, &revoked, &t.CreatedAt)
	if err != nil {
		return model.APIToken{}, err
	}
	if err := json.Unmarshal([]byte(scopes), &t.Scope); err != nil {
		return model.APIToken{}, decodeErr("read a token", "scopes: %v", err)
	}
	t.Owner = opt.FromPtr(owner)
	t.ExpiresAt = opt.FromPtr(expires)
	t.LastUsedAt = opt.FromPtr(lastUsed)
	t.RevokedAt = opt.FromPtr(revoked)
	return t, nil
}

// CreateToken stores a new token.
func (s *Store) CreateToken(ctx context.Context, n NewToken) (model.APIToken, error) {
	const op = "create a token"
	createdAt := s.now()
	scopes, err := json.Marshal(n.Scope)
	if err != nil {
		return model.APIToken{}, dbErr(op, err)
	}
	_, err = exec(ctx, s.db, op, insertToken,
		n.ID.String(), n.OrgID.String(), n.Owner.String(), n.Name, n.Prefix, n.SecretHash, string(scopes),
		n.ExpiresAt.Ptr(), createdAt)
	if err != nil {
		return model.APIToken{}, err
	}
	return model.APIToken{
		ID:         n.ID,
		OrgID:      n.OrgID,
		Owner:      opt.Some(n.Owner),
		Name:       n.Name,
		Prefix:     n.Prefix,
		SecretHash: n.SecretHash,
		Scope:      n.Scope,
		ExpiresAt:  n.ExpiresAt,
		CreatedAt:  createdAt,
	}, nil
}

// FindToken is the token id.
func (s *Store) FindToken(ctx context.Context, id ids.TokenID) (model.APIToken, bool, error) {
	return queryOpt(ctx, s.db, "find a token", selectToken, scanToken, id.String())
}

// ListTokens is owner's personal tokens (not those of CI policies), newest
// first.
func (s *Store) ListTokens(ctx context.Context, owner ids.UserID) ([]model.APIToken, error) {
	return queryAll(ctx, s.db, "list tokens", selectTokensOfOwner, scanToken, owner.String())
}

// RevokeToken revokes one of owner's tokens; false when there was nothing
// to revoke.
func (s *Store) RevokeToken(ctx context.Context, id ids.TokenID, owner ids.UserID) (bool, error) {
	n, err := exec(ctx, s.db, "revoke a token", revokeToken, id.String(), owner.String(), s.now())
	return n > 0, err
}

// RevokeUserTokens revokes every token a user owns in an org (used when a
// member is removed).
func (s *Store) RevokeUserTokens(ctx context.Context, org ids.OrgID, user ids.UserID) (uint64, error) {
	return exec(ctx, s.db, "revoke a user's tokens", revokeUserTokens, org.String(), user.String(), s.now())
}

// TouchToken records that the token was used now.
func (s *Store) TouchToken(ctx context.Context, id ids.TokenID) error {
	_, err := exec(ctx, s.db, "touch a token", touchToken, id.String(), s.now())
	return err
}
